### PR TITLE
Server to work with any spatial dataset CRS

### PR DESCRIPTION
**Pull request by forman**
Note, this PR's client is xcube Viewer PR https://github.com/dcs4cop/xcube-viewer/pull/209

### Enhancements

* `xcube serve` can now serve datasets with arbitrary spatial 
  coordinate reference systems. Before xcube 0.11, datasets where forced
  to have a geographical CRS such as EPSG:4326 or CRS84. 

* `xcube serve` can now provide image tiles for two popular tile grids:
  1. global geographic grid, with 2 x 1 tiles at level zero (the default);
  2. global web mercator grid , with 1 x 1 tiles at level 
     zero ("Google projection", OSM tile grid).
  
  The general form of the xcube tile URL is
       
      /datasets/{ds_id}/vars/{var_name}/tile2/{z}/{y}/{x}
    
  The following query parameters can be used
    
  - `crs`: set to `CRS84` to use the geographical grid (the default),
    or `EPSG:3857` to use the web mercator grid. 
  - `cbar`: color bar name such as `viridis` or `plasma`, 
     see color bar names of matplotlib. Defaults to `bone`.
  - `vmin`: minimum value to be used for color mapping. Defaults to `0`.
  - `vmax`: maximum value to be used for color mapping. Defaults to `1`.
  - `retina`: if set to `1`, tile size will be 512 instead of 256.

* The WMTS provided by `xcube serve` has been reimplemented from scratch.
  It now provides two common tile matrix sets:
  1. `WorldCRS84Quad` global geographic grid, with 2 x 1 tiles at level zero; 
  2. `WorldWebMercatorQuad` global web mercator grid, with 1 x 1 tiles 
     at level zero. 
  
  New RESTful endpoints have been added to reflect this:

      /wmts/1.0.0/{TileMatrixSet}/WMTSCapabilities.xml
      /wmts/1.0.0/tile/{Dataset}/{Variable}/{TileMatrixSet}/{TileMatrix}/{TileRow}/{TileCol}.png
  
  The existing RESTful endpoints now use tile matrix set `WorldCRS84Quad` by default:

      /wmts/1.0.0/WMTSCapabilities.xml
      /wmts/1.0.0/tile/{Dataset}/{Variable}/{TileMatrix}/{TileRow}/{TileCol}.png

  The key-value pair (KVP) endpoint `/wmts/kvp` now recognises the
  `TileMatrixSet` key for the two values described above.

### Other changes

* All components of the `xcube.util.tiledimage` module have been 
  deprecated and are no longer used in xcube.  

* TODO: All components of the `xcube.util.tilegrid` module have been 
  deprecated and are no longer used in xcube. Entirely new implementations 
  are provided in `xcube.core.tilegrid` which are used instead. 

* TODO: All components of the `xcube.core.tile` module have been 
  deprecated and are no longer used in xcube. Entirely new implementations 
  are provided in `xcube.core.tile2` which are used instead. 
  In version xcube 0.12 these will renamed to `xcube.core.tile`.  

Checklist:

* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in `docs/source/*`
* [ ] Changes documented in `CHANGES.md`
* [ ] AppVeyor CI passes
* [ ] Test coverage remains or increases (target 100%)

### CODE REVIEW

File: xcube/core/tilingscheme.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+
+def get_unit_factor(unit_name_from: str, unit_name_to: str) -> float:
+    """
+    Get the factor to convert from one unit into another
+    with units given by *unit_name_from* and *unit_name_to*.
+    """
+    from_meter = _is_meter_unit(unit_name_from)
+    from_degree = _is_degree_unit(unit_name_from)
+    if not from_meter and not from_degree:
+        raise ValueError(f'unsupported units {unit_name_from!r}')
```
Comment (TonioF) [inline xcube/core/tilingscheme.py:398]:
```suggestion
    if not from_meter and not from_degree:
        raise ValueError(f'unsupported units {unit_name_from!r}.'
                         f' Unit must be either meters or degrees.')
```

File: xcube/core/tilingscheme.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+
+
+def _is_meter_unit(unit_name: str) -> bool:
+    return unit_name.lower() in ('m',
+                                 'metre', 'metres',
+                                 'meter', 'meters')
+
+
+def _is_degree_unit(unit_name: str) -> bool:
+    return unit_name.lower() in ('deg',
```
Comment (TonioF) [inline xcube/core/tilingscheme.py:438]:
```suggestion
    return unit_name.lower() in ('°', 'deg',
```

File: xcube/core/timeseries.py [hunk 3]
```diff
@@ -52,55 +58,74 @@
...
+        runtime performance and/or memory consumption.
+    :param cube_asserted: Deprecated and ignored since xcube 0.11.0.
+        No replacement.
     """
-
-    if not cube_asserted:
-        assert_cube(cube)
+    if cube_asserted is not None:
+        warnings.warn('cube_asserted has been deprecated'
+                      ' and will be removed in soon.', DeprecationWarning)
```
Comment (TonioF) [inline xcube/core/timeseries.py:116]:
```suggestion
                      ' and will be removed soon.', DeprecationWarning)
```

File: xcube/util/projcache.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import pyproj
+
+CrsLike = Union[str, pyproj.CRS]
+
+
+class ProjCache:
+    """
+    A cache for pyproj objects
+    that make take considerable time to construct.
```
Comment (TonioF) [inline xcube/util/projcache.py:33]:
```suggestion
    that may take considerable time to construct.
```

File: xcube/util/tilegrid.py [hunk 5]
```diff
@@ -24,11 +24,14 @@
...
 
 import pyproj
+from deprecated import deprecated
 
 from xcube.util.assertions import assert_instance
 from xcube.util.assertions import assert_true
 
 
+@deprecated(version='0.11.0',
+            reason='use xcube.util.tilegrid2.TileGrid2 instead')
```
Comment (TonioF) [inline xcube/util/tilegrid.py:34]: Please do not reference classes that do not exist (yet). This also holds for the other occrurrences of this deprecation message in this module.

**Review by TonioF — CHANGES_REQUESTED** [review]
I'm sorry, but this PR is so big and contains so many different changes that I find it extremely hard to get a handle on it. i tried to look at every removed or newly added line in the code package, I did not look at the tests yet. 
So far, I found some minor things to consider, but most of all, the failing test should be fixed. Also, if there is any chance to break this monster down into several sub-prs (e.g., one for the wmts, one for the new rgb-image-implementation, etc.), this would be extremely helpful. As it stands, it is hard to wrap one's head around it and I can't say that I understand all that's happening here.

File: xcube/util/tilegrid.py (hunk 5 above)
Comment (forman) [reply xcube/util/tilegrid.py:34]: The class referenced here should be the new `xcube.core.tilingscheme.TilingScheme`. Will fix.

**Review by AliceBalfanz — CHANGES_REQUESTED** [review]
Tests are failing as Tonio mentioned. Otherwise I have tested locally and approve :)

**Pull request merged**